enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,275 @@
+via the IPI flow, and existing ignition/MachineConfig APIs are not sufficient
+due to the chicken/egg problem with accessing the MCS.
+
+#### Story 2
+
+As an on-premise IPI user, I wish to use static IPs for my controlplane network,
+for reasons of network ownership or concerns over reliability I can't use DHCP
+and therefore need to provide a static configuration for my primary network.
+
+There is no way to provide [MachineSpecific Configuration in OpenShift](https://github.com/openshift/machine-config-operator/issues/1720) so I am
+forced to use the UPI flow which is less automated and more prone to errors.
+
+### Risks and Mitigations
+
+In some existing workflows, kubernetes-nmstate is used to do network configuration on day-2. Using a different interface for day-1 introduces the potential for mismatches and configuration errors when making day-2 changes. It should be possible to mitigate this by generating the NetworkManager keyfiles using the nmstatecli, which will allow the same nmstate config to be used directly in the day-2 configuration.
+
+An added benefit of writing nmstate configurations instead of NetworkManager
+keyfiles directly is that if/when a day-1 nmstate-based interface becomes
+available the config files can be directly inserted into the new interface.
+However, this is not a complete solution because it still requires manually copying
+configuration to multiple locations which introduces the potential for errors.
+
+## Design Details
+
+In the IPI flow day-1 network configuration is required in 2 different cases:
+
+* Deployment of the controlplane hosts via terraform, using input provided to openshift-install
+* Deployment of compute/worker hosts (during initial deployment and scale-out), via Machine API providers for each platform
+
+In the sections below we will describe the user-facing API that contains network configuration, and the proposed integration for each of these cases.
+
+### User-facing API
+
+RHCOS already provides a mechanism to specify NetworkManager keyfiles during deployment of a new node. We need to expose that functionality during the IPI install process, but preferably using [nmstate](https://nmstate.io) files as the interface for a more user-friendly experience. There are a couple of options on how to do that:
+
+* A new section in install-config.
+* A secret that contains base64-encoded content for the keyfiles.
+
+These are not mutually exclusive. If we implement the install-config option, we will still need to persist the configuration in a secret so it can be used for day-2. In fact, the initial implementation should likely be based on secrets, and then install-config integration can be added on top of that to improve the user experience. This way the install-config work won't block implementation of the feature.
+
+The data provided by the user will need to have the following structure:
+```yaml
+<hostname>: <nmstate configuration>
```

Comment:
I know you mentioned at first it will be per host, but I think that will be really annoying for some users. Assume in my scenario all my hosts are homogeneous and I have 120 of them. That's a giant config. Could we just support applying the same config to all nodes at least initially as well?

Reply:
For the implementation we won't care either way. It's just a question of whether we can give it an interface that makes sense.

Reply:
+1 for default config / templated configs and some sort of node selector mechanism